File: app/src/main/java/com/commit451/gitlab/model/api/MergeRequest.java
package com.commit451.gitlab.model.api;

import android.support.annotation.Nullable;
import android.support.annotation.StringDef;

import com.squareup.moshi.Json;

import org.parceler.Parcel;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.Date;
import java.util.List;

@Parcel
public class MergeRequest {

    public static final String STATE_OPENED = "opened";
    public static final String STATE_MERGED = "merged";
    public static final String STATE_CLOSED = "closed";

    @StringDef({STATE_OPENED, STATE_MERGED, STATE_CLOSED})
    @Retention(RetentionPolicy.SOURCE)
    public @interface State {}

    @Json(name = "id")
    long id;
    @Json(name = "iid")
    long iId;
    @Json(name = "project_id")
    long projectId;
    @Json(name = "title")
    String title;
    @Json(name = "description")
    String description;
    @Json(name = "state")
    @State
    String state;
    @Json(name = "updated_at")
    Date updatedAt;
    @Json(name = "created_at")
    Date createdAt;
    @Json(name = "target_branch")
    String targetBranch;
    @Json(name = "source_branch")
    String sourceBranch;
    @Json(name = "upvotes")
    long upvotes;
    @Json(name = "downvotes")
    long downvotes;
    @Json(name = "author")
    UserBasic author;
    @Json(name = "assignee")
    UserBasic assignee;
    @Json(name = "source_project_id")
    long sourceProjectId;
    @Json(name = "target_project_id")
    long targetProjectId;
    @Json(name = "labels")
    List<String> labels;
    @Json(name = "work_in_progress")
    boolean workInProgress;
    @Json(name = "milestone")
    Milestone milestone;
    @Json(name = "merge_when_build_succeeds")
    boolean mergeWhenBuildSucceeds;
    @Json(name = "merge_status")
    String mergeStatus;
    @Json(name = "changes")
    @Nullable
    List<Diff> changes;

    public MergeRequest() {}

    public long getId() {
        return id;
    }

    public long getIid() {
        return iId;
    }

    public long getProjectId() {
        return projectId;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public @State String getState() {
        return state;
    }

    public Date getUpdatedAt() {
        return updatedAt;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public String getTargetBranch() {
        return targetBranch;
    }

    public String getSourceBranch() {
        return sourceBranch;
    }

    public long getUpvotes() {
        return upvotes;
    }

    public long getDownvotes() {
        return downvotes;
    }

    public UserBasic getAuthor() {
        return author;
    }

    public UserBasic getAssignee() {
        return assignee;
    }

    public long getSourceProjectId() {
        return sourceProjectId;
    }

    public long getTargetProjectId() {
        return targetProjectId;
    }

    public List<String> getLabels() {
        return labels;
    }

    public boolean isWorkInProgress() {
        return workInProgress;
    }

    public Milestone getMilestone() {
        return milestone;
    }

    public boolean isMergeWhenBuildSucceedsEnabled() {
        return mergeWhenBuildSucceeds;
    }

    public String getMergeStatus() {
        return mergeStatus;
    }

    /**
     * Get the changes. Only not null if this merge request was retrieved via {@link com.commit451.gitlab.api.GitLabService#getMergeRequestChanges(long, long)}
     * @return the changes
     */
    @Nullable
    public List<Diff> getChanges() {
        return changes;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof MergeRequest)) {
            return false;
        }

        MergeRequest mergeRequest = (MergeRequest) o;
        return id == mergeRequest.id;
    }

    @Override
    public int hashCode() {
        return (int) (id ^ (id >>> 32));
    }
}

File: app/src/main/java/com/commit451/gitlab/model/api/MergeRequest.kt
package com.commit451.gitlab.model.api

import android.os.Parcelable
import androidx.annotation.StringDef
import com.squareup.moshi.Json
import kotlinx.android.parcel.Parcelize
import java.util.Date

@Parcelize
open class MergeRequest(
    @Json(name = "id")
    var id: Long = 0,
    @Json(name = "iid")
    var iid: Long = 0,
    @Json(name = "project_id")
    var projectId: Long = 0,
    @Json(name = "title")
    var title: String? = null,
    @Json(name = "description")
    var description: String? = null,
    @Json(name = "state")
    @State
    @get:State
    var state: String? = null,
    @Json(name = "updated_at")
    var updatedAt: Date? = null,
    @Json(name = "created_at")
    var createdAt: Date? = null,
    @Json(name = "target_branch")
    var targetBranch: String? = null,
    @Json(name = "source_branch")
    var sourceBranch: String? = null,
    @Json(name = "upvotes")
    var upvotes: Long = 0,
    @Json(name = "downvotes")
    var downvotes: Long = 0,
    @Json(name = "author")
    var author: User? = null,
    @Json(name = "assignee")
    var assignee: User? = null,
    @Json(name = "source_project_id")
    var sourceProjectId: Long? = 0,
    @Json(name = "target_project_id")
    var targetProjectId: Long = 0,
    @Json(name = "labels")
    var labels: List<String>? = null,
    @Json(name = "work_in_progress")
    var isWorkInProgress: Boolean = false,
    @Json(name = "milestone")
    var milestone: Milestone? = null,
    @Json(name = "merge_when_build_succeeds")
    var isMergeWhenBuildSucceedsEnabled: Boolean = false,
    @Json(name = "merge_status")
    var mergeStatus: String? = null,

    /**
     * Get the changes. Only not null if this merge request was retrieved via [com.commit451.gitlab.api.GitLabService.getMergeRequestChanges]
     * @return the changes
     */
    @Json(name = "changes")
    var changes: MutableList<Diff>? = null
) : Parcelable {

    companion object {

        const val STATE_OPENED = "opened"
        const val STATE_MERGED = "merged"
        const val STATE_CLOSED = "closed"
    }

    @StringDef(STATE_OPENED, STATE_MERGED, STATE_CLOSED)
    @Retention(AnnotationRetention.SOURCE)
    annotation class State
}
